File: backend/analysis/redgreen/generate_simulated_db.py
#!/usr/bin/env python3
"""
Generate a simulated.db that mimics a complete REDGREEN experiment.

Reads actual trial simulation_data.json files for frame counts and ground truth,
then simulates realistic key-press sequences for N synthetic participants.

Usage:
    python generate_simulated_db.py [--n-participants 6] [--output simulated.db]
"""

import argparse
import json
import os
import random
import sqlite3
from datetime import datetime, timedelta
from pathlib import Path

# ---------------------------------------------------------------------------
# Paths (relative to this script's location inside backend/analysis/redgreen/)
# ---------------------------------------------------------------------------
SCRIPT_DIR = Path(__file__).resolve().parent
TRIAL_DATA_DIR = SCRIPT_DIR.parent.parent / "trial_data" / "chs_training_zoom"

# V3 familiarization trial order
FTRIAL_NAMES = [
    "T_v3_ball_sensor_red",
    "T_v3_keys",
    "T_v3_area_no_ball",
    "T_v3_green_mid",
    "T_v3_red_mid",
    "T_v3_keyswitch_ball_stable",
    "T_v3_keyswitch_practice_1",
    "T_v3_keyswitch_practice_2",
    "T_v3_occluder_intro",
    "T_v3_occluder_practice",
    "T_v3_before_test",
]


def load_etrial_metadata(trial_data_dir: Path) -> list[dict]:
    """Read E-trial folders and return list of {name, rg_outcome, num_frames}."""
    trials = []
    for entry in sorted(trial_data_dir.iterdir()):
        if not entry.name.startswith("E_"):
            continue
        json_path = entry / "simulation_data.json"
        if not json_path.exists():
            continue
        with open(json_path) as f:
            data = json.load(f)
        trials.append({
            "name": entry.name,
            "rg_outcome": data["rg_outcome"],
            "num_frames": data["num_frames"],
            "fps": data.get("fps", 30),
        })
    return trials


def simulate_keypress_sequence(
    num_frames: int,
    rg_outcome: str,
    accuracy: float,
    switch_rate: float = 0.03,
    rng: random.Random | None = None,
) -> list[dict]:
    """
    Generate a plausible frame-by-frame key-press sequence.

    Parameters
    ----------
    num_frames : int
        Total frames in the trial.
    rg_outcome : str
        Ground truth: 'red' or 'green'.
    accuracy : float
        Probability of pressing the correct key on any given frame (when a key
        is pressed at all). Range [0.3, 0.95] is realistic.
    switch_rate : float
        Per-frame probability of switching response or releasing keys.
    rng : random.Random
        Seeded RNG instance.

    Returns
    -------
    list of dict
        Each dict has keys: frame, f_pressed, j_pressed, relative_time_ms.
    """
    if rng is None:
        rng = random.Random()

    # Correct key: F=green, J=red
    correct_f = 1 if rg_outcome == "green" else 0
    correct_j = 1 if rg_outcome == "red" else 0

    frames = []
    # Start with no key pressed (first ~10% of frames tend to be uncertain)
    f_state, j_state = 0, 0
    commit_frame = int(num_frames * rng.uniform(0.05, 0.25))

    for i in range(num_frames):
        # Phase 1: initial uncertainty (no key pressed)
        if i < commit_frame:
            # Small chance of early exploratory press
            if rng.random() < 0.05:
                if rng.random() < 0.5:
                    f_state, j_state = 1, 0
                else:
                    f_state, j_state = 0, 1
            else:
                f_state, j_state = 0, 0
        else:
            # Phase 2: active responding — mostly hold one key, occasionally switch
            if rng.random() < switch_rate:
                # Switch or release
                roll = rng.random()
                if roll < 0.15:
                    # Both released briefly
                    f_state, j_state = 0, 0
                elif roll < 0.3:
                    # Switch to incorrect
                    f_state = 1 - correct_f
                    j_state = 1 - correct_j
                else:
                    # Switch to correct
                    f_state = correct_f
                    j_state = correct_j
            elif f_state == 0 and j_state == 0:
                # Re-engage: press correct key with probability = accuracy
                if rng.random() < accuracy:
                    f_state, j_state = correct_f, correct_j
                else:
                    f_state, j_state = 1 - correct_f, 1 - correct_j

        # Timing: ~33ms per frame with jitter
        relative_time_ms = round(i * (1000.0 / 30) + rng.gauss(0, 2), 1)
        if i == 0:
            relative_time_ms = 0.0

        frames.append({
            "frame": i,
            "f_pressed": f_state,
            "j_pressed": j_state,
            "relative_time_ms": max(0.0, relative_time_ms),
        })

    return frames


def compute_score(keystates: list[dict], rg_outcome: str) -> float:
    """Compute trial score using the backend formula."""
    correct_key_f = 1 if rg_outcome == "green" else 0
    correct_key_j = 1 if rg_outcome == "red" else 0

    correct = 0
    incorrect = 0
    for ks in keystates:
        f, j = ks["f_pressed"], ks["j_pressed"]
        if f == correct_key_f and j == correct_key_j and (f != j):
            correct += 1
        elif f != correct_key_f and j != correct_key_j and (f != j):
            incorrect += 1
        # Both pressed or neither: not counted

    total = len(keystates)
    if total == 0:
        return 20.0
    return 20.0 + 100.0 * ((correct - incorrect) / total)


def create_db(db_path: str, n_participants: int, seed: int = 42):
    """Generate the simulated database."""
    rng = random.Random(seed)
    etrial_meta = load_etrial_metadata(TRIAL_DATA_DIR)

    if not etrial_meta:
        raise FileNotFoundError(
            f"No E-trial folders found in {TRIAL_DATA_DIR}. "
            "Ensure trial_data/chs_training_zoom/E_*/simulation_data.json exist."
        )

    print(f"Found {len(etrial_meta)} E-trials")
    print(f"Generating data for {n_participants} simulated participants")

    conn = sqlite3.connect(db_path)
    c = conn.cursor()

    # Create tables matching the backend schema
    c.executescript("""
        CREATE TABLE IF NOT EXISTS redgreen_session (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            randomized_profile_id INTEGER,
            start_time DATETIME,
            end_time DATETIME,
            prolific_pid VARCHAR(100),
            study_id VARCHAR(100),
            prolific_session_id VARCHAR(100),
            average_score FLOAT,
            time_taken FLOAT,
            randomized_trial_order JSON,
            ignore_data BOOLEAN,
            completed BOOLEAN,
            has_timed_out BOOLEAN,
            experiment_name VARCHAR(100)
        );

        CREATE TABLE IF NOT EXISTS trial (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            session_id INTEGER NOT NULL,
            start_time DATETIME,
            end_time DATETIME,
            trial_type VARCHAR(20),
            trial_index INTEGER,
            global_trial_name VARCHAR(100),
            counterbalance BOOLEAN,
            score FLOAT,
            completed BOOLEAN,
            first_frame_utc DATETIME,
            last_frame_utc DATETIME,
            FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
        );

        CREATE TABLE IF NOT EXISTS keystate (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            trial_id INTEGER NOT NULL,
            frame INTEGER,
            f_pressed BOOLEAN,
            j_pressed BOOLEAN,
            session_id INTEGER NOT NULL,
            relative_time_ms FLOAT,
            FOREIGN KEY (trial_id) REFERENCES trial(id),
            FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
        );

        CREATE TABLE IF NOT EXISTS config (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            session_id INTEGER NOT NULL,
            config_data BLOB NOT NULL,
            FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
        );
    """)

    base_time = datetime(2026, 3, 1, 10, 0, 0)

    for p in range(n_participants):
        # Each participant has a different accuracy level
        participant_accuracy = rng.uniform(0.4, 0.9)
        profile_id = p + 1
        pid = f"sim_participant_{p+1:03d}"

        # Randomize E-trial order for this participant
        etrial_order = [t["name"] for t in etrial_meta]
        rng.shuffle(etrial_order)

        session_start = base_time + timedelta(hours=p * 2, minutes=rng.randint(0, 30))

        # Insert session (we'll update average_score and end_time later)
        c.execute(
            """INSERT INTO redgreen_session
               (randomized_profile_id, start_time, prolific_pid, study_id,
                prolific_session_id, randomized_trial_order, ignore_data,
                completed, has_timed_out, experiment_name)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                profile_id,
                session_start.isoformat(),
                pid,
                "simulated_study",
                f"sim_session_{pid}",
                json.dumps(etrial_order),
                0,  # ignore_data
                1,  # completed
                0,  # has_timed_out
                "redgreen",
            ),
        )
        session_id = c.lastrowid
        trial_time = session_start + timedelta(seconds=30)

        # -- Familiarization trials (no keystate data, not completed) --
        for fi, fname in enumerate(FTRIAL_NAMES):
            c.execute(
                """INSERT INTO trial
                   (session_id, start_time, trial_type, trial_index,
                    global_trial_name, counterbalance, completed)
                   VALUES (?, ?, ?, ?, ?, ?, ?)""",
                (session_id, trial_time.isoformat(), "ftrial", fi, f"F{fi+1}", 0, 0),
            )
            trial_time += timedelta(seconds=rng.uniform(15, 45))

        # -- Experimental trials --
        trial_scores = []
        etrial_meta_map = {t["name"]: t for t in etrial_meta}

        for ei, trial_name in enumerate(etrial_order):
            meta = etrial_meta_map[trial_name]
            num_frames = meta["num_frames"]
            rg_outcome = meta["rg_outcome"]

            trial_start = trial_time
            trial_duration = num_frames / 30.0

            # Generate key-press sequence
            keystates = simulate_keypress_sequence(
                num_frames, rg_outcome, participant_accuracy, rng=rng
            )
            score = compute_score(keystates, rg_outcome)
            trial_scores.append(score)

            first_frame_utc = trial_start
            last_frame_utc = trial_start + timedelta(seconds=trial_duration)

            c.execute(
                """INSERT INTO trial
                   (session_id, start_time, end_time, trial_type, trial_index,
                    global_trial_name, counterbalance, score, completed,
                    first_frame_utc, last_frame_utc)
                   VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
                (
                    session_id,
                    trial_start.isoformat(),
                    last_frame_utc.isoformat(),
                    "trial",
                    ei,
                    trial_name,
                    0,  # counterbalance off
                    score,
                    1,  # completed
                    first_frame_utc.isoformat(),
                    last_frame_utc.isoformat(),
                ),
            )
            trial_id = c.lastrowid

            # Insert keystate rows
            for ks in keystates:
                c.execute(
                    """INSERT INTO keystate
                       (trial_id, frame, f_pressed, j_pressed, session_id, relative_time_ms)
                       VALUES (?, ?, ?, ?, ?, ?)""",
                    (
                        trial_id,
                        ks["frame"],
                        ks["f_pressed"],
                        ks["j_pressed"],
                        session_id,
                        ks["relative_time_ms"],
                    ),
                )

            trial_time = last_frame_utc + timedelta(seconds=rng.uniform(3, 8))

        # Update session with final stats
        session_end = trial_time
        avg_score = sum(trial_scores) / len(trial_scores) if trial_scores else 0
        time_taken = (session_end - session_start).total_seconds()

        c.execute(
            """UPDATE redgreen_session
               SET average_score = ?, time_taken = ?, end_time = ?
               WHERE id = ?""",
            (avg_score, time_taken, session_end.isoformat(), session_id),
        )

        print(f"  Participant {p+1}/{n_participants}: {pid}, "
              f"accuracy={participant_accuracy:.2f}, "
              f"avg_score={avg_score:.1f}, "
              f"{len(etrial_order)} trials")

    conn.commit()

    # Summary
    c.execute("SELECT COUNT(*) FROM redgreen_session")
    n_sess = c.fetchone()[0]
    c.execute("SELECT COUNT(*) FROM trial WHERE trial_type='trial' AND completed=1")
    n_trials = c.fetchone()[0]
    c.execute("SELECT COUNT(*) FROM keystate")
    n_ks = c.fetchone()[0]

    print(f"\nSimulated DB written to: {db_path}")
    print(f"  Sessions: {n_sess}")
    print(f"  Completed E-trials: {n_trials}")
    print(f"  Keystate rows: {n_ks}")

    conn.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate simulated REDGREEN DB")
    parser.add_argument("--n-participants", type=int, default=6,
                        help="Number of simulated participants (default: 6)")
    parser.add_argument("--output", type=str, default=None,
                        help="Output DB path (default: backend/analysis/redgreen/simulated.db)")
    parser.add_argument("--seed", type=int, default=42, help="Random seed")
    args = parser.parse_args()

    if args.output is None:
        args.output = str(SCRIPT_DIR / "simulated.db")

    create_db(args.output, args.n_participants, args.seed)

File: backend/analysis/redgreen/generate_simulated_db.R
# REDGREEN Experiment — Generate Simulated Database
# ==================================================
#
# R equivalent of generate_simulated_db.py. Uses the sqlite3 CLI to write the DB.
#
# Reads actual trial simulation_data.json files for frame counts and ground truth,
# then simulates realistic key-press sequences for N synthetic participants.
#
# Usage:
#   Rscript generate_simulated_db.R [--n-participants 6] [--output simulated.db] [--seed 42]
#
# Required packages: jsonlite
# Required system tool: sqlite3 CLI
# ---------------------------------------------------------------------------

suppressPackageStartupMessages({
  library(jsonlite)
})


# =========================================================================
# Paths (relative to this script)
# =========================================================================

get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg[1]))))
  }
  getwd()
}

SCRIPT_DIR <- get_script_dir()
TRIAL_DATA_DIR <- file.path(SCRIPT_DIR, "..", "..", "trial_data", "chs_training_zoom")


# =========================================================================
# Helper: execute SQL via sqlite3 CLI
# =========================================================================

sqlite3_exec <- function(db_path, sql) {
  # Pipe SQL to sqlite3 stdin to avoid quoting issues.
  tmp <- tempfile(fileext = ".sql")
  on.exit(unlink(tmp))
  writeLines(sql, tmp)
  system2("sqlite3", args = shQuote(db_path), stdin = tmp,
          stdout = TRUE, stderr = TRUE)
}

sqlite3_query_plain <- function(db_path, sql) {
  # Run a SELECT and return raw lines (no header, no CSV quoting).
  tmp <- tempfile(fileext = ".sql")
  on.exit(unlink(tmp))
  writeLines(sql, tmp)
  result <- system2("sqlite3", args = shQuote(db_path), stdin = tmp,
                    stdout = TRUE, stderr = FALSE)
  result[nzchar(result)]
}


# =========================================================================
# Load E-trial metadata from simulation JSONs
# =========================================================================

load_etrial_metadata <- function(trial_data_dir) {
  entries <- list.dirs(trial_data_dir, full.names = FALSE, recursive = FALSE)
  e_folders <- sort(entries[startsWith(entries, "E")])

  results <- lapply(e_folders, function(folder) {
    json_path <- file.path(trial_data_dir, folder, "simulation_data.json")
    if (!file.exists(json_path)) return(NULL)
    d <- fromJSON(json_path)
    data.frame(
      name = folder,
      rg_outcome = d$rg_outcome,
      num_frames = as.integer(d$num_frames),
      fps = as.integer(d$fps %||% 30L),
      stringsAsFactors = FALSE
    )
  })

  do.call(rbind, Filter(Negate(is.null), results))
}


# =========================================================================
# Simulate a key-press sequence for one trial
# =========================================================================

simulate_keypress_sequence <- function(num_frames, rg_outcome, accuracy,
                                       switch_rate = 0.03) {
  # Correct key: F=green, J=red
  correct_f <- if (rg_outcome == "green") 1L else 0L
  correct_j <- if (rg_outcome == "red")   1L else 0L

  f_state <- 0L
  j_state <- 0L
  commit_frame <- as.integer(num_frames * runif(1, 0.05, 0.25))

  frames <- vector("list", num_frames)

  for (i in seq_len(num_frames)) {
    idx <- i - 1L  # 0-based frame

    if (idx < commit_frame) {
      # Phase 1: initial uncertainty
      if (runif(1) < 0.05) {
        if (runif(1) < 0.5) {
          f_state <- 1L; j_state <- 0L
        } else {
          f_state <- 0L; j_state <- 1L
        }
      } else {
        f_state <- 0L; j_state <- 0L
      }
    } else {
      # Phase 2: active responding
      if (runif(1) < switch_rate) {
        roll <- runif(1)
        if (roll < 0.15) {
          f_state <- 0L; j_state <- 0L
        } else if (roll < 0.3) {
          f_state <- 1L - correct_f; j_state <- 1L - correct_j
        } else {
          f_state <- correct_f; j_state <- correct_j
        }
      } else if (f_state == 0L && j_state == 0L) {
        if (runif(1) < accuracy) {
          f_state <- correct_f; j_state <- correct_j
        } else {
          f_state <- 1L - correct_f; j_state <- 1L - correct_j
        }
      }
    }

    # Timing: ~33ms per frame with jitter
    rel_time <- if (idx == 0L) 0.0 else max(0.0, round(idx * (1000.0 / 30) + rnorm(1, 0, 2), 1))

    frames[[i]] <- list(frame = idx, f_pressed = f_state, j_pressed = j_state,
                        relative_time_ms = rel_time)
  }

  frames
}


# =========================================================================
# Compute trial score (backend formula)
# =========================================================================

compute_score <- function(keystates, rg_outcome) {
  correct_f <- if (rg_outcome == "green") 1L else 0L
  correct_j <- if (rg_outcome == "red")   1L else 0L

  correct <- 0L
  incorrect <- 0L

  for (ks in keystates) {
    f <- ks$f_pressed; j <- ks$j_pressed
    if (f == correct_f && j == correct_j && f != j) {
      correct <- correct + 1L
    } else if (f != correct_f && j != correct_j && f != j) {
      incorrect <- incorrect + 1L
    }
  }

  total <- length(keystates)
  if (total == 0L) return(20.0)
  20.0 + 100.0 * ((correct - incorrect) / total)
}


# =========================================================================
# SQL escaping helper
# =========================================================================

sql_quote <- function(x) {
  if (is.null(x) || is.na(x)) return("NULL")
  if (is.numeric(x)) return(as.character(x))
  # Escape single quotes by doubling them.
  paste0("'", gsub("'", "''", as.character(x)), "'")
}


# =========================================================================
# Main: generate the simulated database
# =========================================================================

create_db <- function(db_path, n_participants = 6L, seed = 42L,
                      trial_data_dir = TRIAL_DATA_DIR) {

  set.seed(seed)

  etrial_meta <- load_etrial_metadata(trial_data_dir)
  if (is.null(etrial_meta) || nrow(etrial_meta) == 0) {
    stop(sprintf("No E-trial folders found in %s", trial_data_dir))
  }

  cat(sprintf("Found %d E-trials\n", nrow(etrial_meta)))
  cat(sprintf("Generating data for %d simulated participants\n", n_participants))

  # Remove existing DB if present.
  if (file.exists(db_path)) unlink(db_path)

  # V3 familiarization trial names.
  ftrial_names <- c(
    "T_v3_ball_sensor_red", "T_v3_keys", "T_v3_area_no_ball",
    "T_v3_green_mid", "T_v3_red_mid", "T_v3_keyswitch_ball_stable",
    "T_v3_keyswitch_practice_1", "T_v3_keyswitch_practice_2",
    "T_v3_occluder_intro", "T_v3_occluder_practice", "T_v3_before_test"
  )

  # --- Create schema ---
  sqlite3_exec(db_path, "
    CREATE TABLE redgreen_session (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      randomized_profile_id INTEGER,
      start_time DATETIME,
      end_time DATETIME,
      prolific_pid VARCHAR(100),
      study_id VARCHAR(100),
      prolific_session_id VARCHAR(100),
      average_score FLOAT,
      time_taken FLOAT,
      randomized_trial_order JSON,
      ignore_data BOOLEAN,
      completed BOOLEAN,
      has_timed_out BOOLEAN,
      experiment_name VARCHAR(100)
    );

    CREATE TABLE trial (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      session_id INTEGER NOT NULL,
      start_time DATETIME,
      end_time DATETIME,
      trial_type VARCHAR(20),
      trial_index INTEGER,
      global_trial_name VARCHAR(100),
      counterbalance BOOLEAN,
      score FLOAT,
      completed BOOLEAN,
      first_frame_utc DATETIME,
      last_frame_utc DATETIME,
      FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
    );

    CREATE TABLE keystate (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      trial_id INTEGER NOT NULL,
      frame INTEGER,
      f_pressed BOOLEAN,
      j_pressed BOOLEAN,
      session_id INTEGER NOT NULL,
      relative_time_ms FLOAT,
      FOREIGN KEY (trial_id) REFERENCES trial(id),
      FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
    );

    CREATE TABLE config (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      session_id INTEGER NOT NULL,
      config_data BLOB NOT NULL,
      FOREIGN KEY (session_id) REFERENCES redgreen_session(id)
    );
  ")

  base_time <- as.POSIXct("2026-03-01 10:00:00", tz = "UTC")
  fmt <- "%Y-%m-%d %H:%M:%OS6"

  for (p in seq_len(n_participants)) {
    accuracy <- runif(1, 0.4, 0.9)
    profile_id <- p
    pid <- sprintf("sim_participant_%03d", p)

    # Randomize E-trial order.
    etrial_order <- sample(etrial_meta$name)

    session_start <- base_time + (p - 1) * 7200 + sample(0:1800, 1)

    # --- Insert session ---
    order_json <- toJSON(etrial_order, auto_unbox = TRUE)
    session_sql <- sprintf(
      "INSERT INTO redgreen_session
       (randomized_profile_id, start_time, prolific_pid, study_id,
        prolific_session_id, randomized_trial_order, ignore_data,
        completed, has_timed_out, experiment_name)
       VALUES (%d, %s, %s, %s, %s, %s, 0, 1, 0, 'redgreen');",
      profile_id,
      sql_quote(format(session_start, fmt)),
      sql_quote(pid),
      sql_quote("simulated_study"),
      sql_quote(paste0("sim_session_", pid)),
      sql_quote(as.character(order_json))
    )
    sqlite3_exec(db_path, session_sql)

    # Get session_id (auto-increment = p).
    session_id <- p
    trial_time <- session_start + 30

    # --- Familiarization trials (no keystate, not completed) ---
    ftrial_sqls <- character(length(ftrial_names))
    for (fi in seq_along(ftrial_names)) {
      ftrial_sqls[fi] <- sprintf(
        "INSERT INTO trial (session_id, start_time, trial_type, trial_index,
         global_trial_name, counterbalance, completed)
         VALUES (%d, %s, 'ftrial', %d, %s, 0, 0);",
        session_id, sql_quote(format(trial_time, fmt)),
        fi - 1L, sql_quote(sprintf("F%d", fi))
      )
      trial_time <- trial_time + runif(1, 15, 45)
    }
    sqlite3_exec(db_path, paste(ftrial_sqls, collapse = "\n"))

    # --- Experimental trials ---
    trial_scores <- numeric(length(etrial_order))

    # Build a lookup for trial metadata.
    meta_lookup <- setNames(
      split(etrial_meta, seq_len(nrow(etrial_meta))),
      etrial_meta$name
    )

    # Collect all SQL for this participant in batches.
    all_trial_sql <- character(length(etrial_order))
    all_ks_sql <- list()

    for (ei in seq_along(etrial_order)) {
      tname <- etrial_order[ei]
      meta <- meta_lookup[[tname]]
      num_frames <- meta$num_frames
      rg_outcome <- meta$rg_outcome

      trial_start <- trial_time
      trial_duration <- num_frames / 30.0

      # Simulate key presses.
      keystates <- simulate_keypress_sequence(num_frames, rg_outcome, accuracy)
      score <- compute_score(keystates, rg_outcome)
      trial_scores[ei] <- score

      first_frame_utc <- trial_start
      last_frame_utc <- trial_start + trial_duration

      # trial_id will be: length(ftrial_names) + (p-1)*total_trials_per_session + ei
      # But we need actual auto-increment IDs. We'll query after insert.
      all_trial_sql[ei] <- sprintf(
        "INSERT INTO trial
         (session_id, start_time, end_time, trial_type, trial_index,
          global_trial_name, counterbalance, score, completed,
          first_frame_utc, last_frame_utc)
         VALUES (%d, %s, %s, 'trial', %d, %s, 0, %f, 1, %s, %s);",
        session_id,
        sql_quote(format(trial_start, fmt)),
        sql_quote(format(last_frame_utc, fmt)),
        ei - 1L,
        sql_quote(tname),
        score,
        sql_quote(format(first_frame_utc, fmt)),
        sql_quote(format(last_frame_utc, fmt))
      )

      # Store keystates with a placeholder; we need trial_id after insert.
      all_ks_sql[[ei]] <- keystates

      trial_time <- last_frame_utc + runif(1, 3, 8)
    }

    # Insert all trials for this participant.
    sqlite3_exec(db_path, paste(all_trial_sql, collapse = "\n"))

    # Now get the trial IDs we just inserted (E-trials for this session).
    trial_id_lines <- sqlite3_query_plain(db_path, sprintf(
      "SELECT id FROM trial WHERE session_id = %d AND trial_type = 'trial' ORDER BY id;",
      session_id
    ))
    trial_ids <- as.integer(trial_id_lines)

    if (length(trial_ids) != length(etrial_order)) {
      stop(sprintf("Expected %d trial IDs, got %d", length(etrial_order), length(trial_ids)))
    }

    # Insert keystates in batches (large INSERT with multiple VALUES).
    cat(sprintf("  Participant %d/%d: %s, accuracy=%.2f, inserting keystates...",
                p, n_participants, pid, accuracy))

    batch_size <- 500L
    ks_values <- character(0)

    for (ei in seq_along(all_ks_sql)) {
      tid <- trial_ids[ei]
      for (ks in all_ks_sql[[ei]]) {
        ks_values <- c(ks_values, sprintf(
          "(%d, %d, %d, %d, %d, %f)",
          tid, ks$frame, ks$f_pressed, ks$j_pressed, session_id, ks$relative_time_ms
        ))

        if (length(ks_values) >= batch_size) {
          sqlite3_exec(db_path, sprintf(
            "INSERT INTO keystate (trial_id, frame, f_pressed, j_pressed, session_id, relative_time_ms) VALUES %s;",
            paste(ks_values, collapse = ",\n")
          ))
          ks_values <- character(0)
        }
      }
    }
    # Flush remaining.
    if (length(ks_values) > 0) {
      sqlite3_exec(db_path, sprintf(
        "INSERT INTO keystate (trial_id, frame, f_pressed, j_pressed, session_id, relative_time_ms) VALUES %s;",
        paste(ks_values, collapse = ",\n")
      ))
    }

    # Update session with final stats.
    session_end <- trial_time
    avg_score <- mean(trial_scores)
    time_taken <- as.numeric(difftime(session_end, session_start, units = "secs"))

    sqlite3_exec(db_path, sprintf(
      "UPDATE redgreen_session SET average_score = %f, time_taken = %f, end_time = %s WHERE id = %d;",
      avg_score, time_taken, sql_quote(format(session_end, fmt)), session_id
    ))

    cat(sprintf(" avg_score=%.1f\n", avg_score))
  }

  # Summary.
  n_sess <- sqlite3_query_plain(db_path, "SELECT COUNT(*) FROM redgreen_session;")
  n_trials <- sqlite3_query_plain(db_path,
    "SELECT COUNT(*) FROM trial WHERE trial_type='trial' AND completed=1;")
  n_ks <- sqlite3_query_plain(db_path, "SELECT COUNT(*) FROM keystate;")

  cat(sprintf("\nSimulated DB written to: %s\n", db_path))
  cat(sprintf("  Sessions: %s\n  Completed E-trials: %s\n  Keystate rows: %s\n",
              n_sess, n_trials, n_ks))
}


# =========================================================================
# CLI argument parsing
# =========================================================================

if (!interactive()) {
  args <- commandArgs(trailingOnly = TRUE)

  n_participants <- 6L
  output <- file.path(SCRIPT_DIR, "simulated.db")
  seed <- 42L

  i <- 1L
  while (i <= length(args)) {
    if (args[i] == "--n-participants" && i < length(args)) {
      n_participants <- as.integer(args[i + 1L])
      i <- i + 2L
    } else if (args[i] == "--output" && i < length(args)) {
      output <- args[i + 1L]
      i <- i + 2L
    } else if (args[i] == "--seed" && i < length(args)) {
      seed <- as.integer(args[i + 1L])
      i <- i + 2L
    } else {
      cat(sprintf("Unknown argument: %s\n", args[i]))
      cat("Usage: Rscript generate_simulated_db.R [--n-participants N] [--output PATH] [--seed S]\n")
      quit(status = 1)
    }
  }

  create_db(output, n_participants, seed)
}
